### PR TITLE
5pm-c-Num full courses by department

### LINKED ISSUE #43: Number of full courses in current department
# User Story

As a user I can find the number of courses that are full, given a start and end quarter, and a department so that users can see how popular a class was in the past in order to assist in choosing their future courses.

# Acceptance Criteria

- [x] In the Statistics dropdown menu, there should be an entry that links to a page that allows you to search for the number of full courses.
- [x] There is a form on the full courses page that lets you select a department and quarter range.
- [x] After submitting the form, it should display the number of full courses.
...

# Implementation Todos

Front end:

- [x] Add a form with a drop down selection of start quarter, end quarter, and department
- [x] Add a service that allows the form to fetch the given division occupancy and set the course JSON in order for it to display below
- [x] Implement a counter that shows the number of full courses the query to MongoDB found for the selected department and quarter range (Possibly show a table of all the full courses?)
...

Back end:

- [x] Add a endpoint to the StatisticsController file that will service fetch requests from the frontend
- [x] In the query, only get courses that have all their sections full or have total capacity full.
- [x] In the StatisticsController file, use MongoDB and aggregate the quarter/department data and return as JSON.
...

Testing:

- [x] Front end tests pass and there is adequate coverage
- [x] Back end tests pass and there is adequate coverage
...

### PR DESCRIPTION
**Pull request by zminshall**
This PR implements the "Full Courses By Department" statistic page, where a user can select a start and end quarter along with the department to get the number of full courses and a table of the courses in that department over that quarter range that are full. Closes #43

### CODE REVIEW

**Review by hdinh77 — APPROVED** [review]
full coverage tests run on local passed, no bugs/errors, no unnecessary lines

File: javascript/src/main/components/Statistics/FullCoursesForm.js
Comment (jvoucsb) [inline javascript/src/main/components/Statistics/FullCoursesForm.js:16]: Seems to be unnecessary console.log statement for debugging. Consider removing

**Review by jvoucsb — COMMENTED** [review]
Looks good to me. I tested the branch and found no bugs in functionality. Consider this one nitpicky change and after I'd approve it.

File: javascript/src/main/components/Statistics/FullCoursesForm.js
Comment (zminshall) [reply javascript/src/main/components/Statistics/FullCoursesForm.js]: Fixed!

**Review by hdinh77 — APPROVED** [review]
looks good-- all unnecessary whitespace and console.log statements removed

**Review by jvoucsb — DISMISSED** [review]
Looks good to me

File: javascript/src/main/components/Statistics/DepartmentFormSelect.js
Comment (pconrad) [inline javascript/src/main/components/Statistics/DepartmentFormSelect.js:69]: I added an issue for the W21 team to replace this with an API call.  It's fine that it's hard coded for now.

File: javascript/src/main/components/Statistics/FullCourseTable.js
Comment (pconrad) [inline javascript/src/main/components/Statistics/FullCourseTable.js:12]: Instead of
```
const d2d4d7 = {
  backgroundColor: '#d2d4d7'
}
``` 

I wonder if it would be better to give this color a semantically meaningful name?  How is it used in the code?  

e.g. 
```
const sectionColor = ...
const lectureColor = ...
const fullSectionColor = ...
```

File: javascript/src/main/components/Statistics/FullCourseTable.js
Comment (pconrad) [inline javascript/src/main/components/Statistics/FullCourseTable.js:19]: Same... maybe 
```
const oddRows = '#e9ecef';
const evenRows = '#edf0f2';
```
And put all of these colors into a single JavaScript file that is imported, and then used everywhere these colors are needed in the app?

File: javascript/src/main/components/Statistics/FullCourseTable.js
Comment (pconrad) [inline javascript/src/main/components/Statistics/FullCourseTable.js:41]: Can we dry this up, by specifying

```
const footerStyleColor = `d2d4d7`; 
```

and then using the variable value?

**Review by pconrad — DISMISSED** [review]
I have a few suggestions, but these are things that don't need to hold up the PR

**Comment by pconrad** [comment]
I think this one is likely ready to merge.  I will test it in the morning on qa, and then it is probably good to go

File: javascript/src/main/components/Statistics/FullCourseTable.js
Comment (zminshall) [reply javascript/src/main/components/Statistics/FullCourseTable.js:74]: Updated!

File: javascript/src/main/components/Statistics/FullCourseTable.js
Comment (zminshall) [reply javascript/src/main/components/Statistics/FullCourseTable.js:43]: Updated!

File: javascript/src/main/components/Statistics/FullCourseTable.js
Comment (zminshall) [reply javascript/src/main/components/Statistics/FullCourseTable.js:35]: Updated!

**Review by jvoucsb — APPROVED** [review]
looks good to me

**Review by jvoucsb — APPROVED** [review]
Looks good to me. Deployed to make sure the feature worked.

**Review by JacquelineMai — APPROVED** [review]

**Pull request merged**
